### PR TITLE
feat(typescript): support inlined types (and initialize an AST)

### PR DESCRIPTION
**Pull request by dsinghvi**
## Summary by CodeRabbit

- **New Features**
	- Introduced new classes, including `AstNode`, `Writer`, `DocString`, and `TypescriptFile`, enhancing TypeScript code generation capabilities.
	- Added the `respectInlinedTypes` configuration option to allow more customization in the SDK generation process.
	- Expanded the `InternalType` definition to include additional types for improved type handling.

- **Bug Fixes**
	- Updated dependency versions for `@fern-fern/ir-sdk` to ensure compatibility and access to new features.
	- Adjusted import paths and property declaration syntax for improved clarity.

- **Documentation**
	- Enhanced unit tests for the `Writer` class to validate output consistency across various scenarios.

- **Refactor**
	- Improved code organization and handling of type generation across multiple classes.
	- Simplified constructor parameters in the `SdkGenerator` and `SdkContextImpl` classes.

### CODE REVIEW

File: generators/typescript/codegen/src/ast/TypescriptFile.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    public write(writer: Writer): void {
+        for (const element of this.elements) {
+            writer.writeNode(element);
+            writer.writeLine();
+            writer.writeLine();
+            writer.writeLine();
+        }
+    }
+}
```
Comment (dsinghvi) [inline generators/typescript/codegen/src/ast/TypescriptFile.ts:30]: going to delete this

File: generators/typescript/codegen/src/ast/Namespace.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    public write(writer: Writer): void {
+        if (this.args.export) {
+            writer.write("export ");
+        }
+
+        let setAmbience = false;
+        if (!writer.isAmbient) {
+            writer.write("declare ");
+            writer.setAmbient(true);
```
Comment (amckinney) [inline generators/typescript/codegen/src/ast/Namespace.ts:40]: nit: I like the `startControlFlow` and `endControlFlow` naming that we have going. Maybe we do `writer.startAmbient()` and `writer.stopAmbient()`? Alternatively we just modify the public `isAmbient` property directly.

File: generators/typescript/codegen/src/ast/Reference.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        type: "module";
+        /* The package or path to import from  */
+        source: string;
+        /* The module to import from */
+        module: string;
+        /**
+         * The path to access the reference from the module.
+         * [module, ...name].join(".") is the reference
+         */
+        name: string[];
```
Comment (amckinney) [inline generators/typescript/codegen/src/ast/Reference.ts:26]: Should this be named `path`? And should we separate the final element into its own property named `name`? So the reference would effectively be:

```
[module, ...path, name].join(".")
```

If this is too much of a pain, happy to leave this as-is named `name`.

File: generators/typescript/codegen/src/ast/Type.ts [hunk 4]
```diff
@@ -18,8 +92,65 @@ export class Type extends AstNode {
...
+                writer.write("unknown");
+                break;
+            case "union": {
+                const numVariants = this.internalType.variants.length;
+                this.internalType.variants.forEach((variant, idx) => {
+                    writer.writeNode(variant);
+                    if (idx < numVariants - 1) {
+                        writer.write(" | ");
+                    }
+                });
```
Comment (amckinney) [inline generators/typescript/codegen/src/ast/Type.ts:149]: nit: similar to the other comment:
```suggestion
                this.internalType.variants.forEach((variant, idx) => {
                    if (idx > 0) {
                        writer.write(" | ");
                    }
                    writer.writeNode(variant);
                });
```

File: generators/typescript/codegen/src/ast/__test__/__snapshots__/TypescriptFile.test.ts.snap [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+"
+`;
+
+exports[`TypescriptFile > toStringFormatted > should generate an empty TypeScript file 1`] = `""`;
+
+exports[`TypescriptFile > toStringFormatted > should maintain the correct order of added elements 1`] = `
+"interface Interface1 {}
+declare namespace Namespace1 {}
+interface Interface2 {}
+declare namespace Namespace2 {}
```
Comment (amckinney) [inline generators/typescript/codegen/src/ast/__test__/__snapshots__/TypescriptFile.test.ts.snap:64]: Should these have newlines between them, or is this intentional?

File: generators/typescript/codegen/src/context/BaseTypescriptGeneratorContext.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+    public getReferenceToNamedType(typeId: TypeId): ts.Reference {
+        const typeDeclaration = this.getTypeDeclarationOrThrow(typeId);
+        return Reference.module({
+            module: this.getNamespaceExport(),
+            name: [
+                ...typeDeclaration.name.fernFilepath.packagePath.map((path) => path.pascalCase.safeName),
+                typeDeclaration.name.name.pascalCase.safeName
+            ],
+            source: `${typeDeclaration.name.fernFilepath.allParts.map(() => "../../../").join("")}`
```
Comment (amckinney) [inline generators/typescript/codegen/src/context/BaseTypescriptGeneratorContext.ts:51]: Is `../../../` reliable at this point? How do we know if we're _always_ that far nested within?

File: generators/typescript/codegen/src/ast/__test__/__snapshots__/TypescriptFile.test.ts.snap (hunk 5 above)
Comment (dsinghvi) [reply generators/typescript/codegen/src/ast/__test__/__snapshots__/TypescriptFile.test.ts.snap:64]: intentional (how prettier formats)

File: generators/typescript/codegen/src/context/BaseTypescriptGeneratorContext.ts (hunk 6 above)
Comment (dsinghvi) [reply generators/typescript/codegen/src/context/BaseTypescriptGeneratorContext.ts:51]: Yup -- every fern file path part is 3 directories nested for dumb legacy reasons

File: generators/typescript/codegen/src/ast/Reference.ts (hunk 3 above)
Comment (dsinghvi) [reply generators/typescript/codegen/src/ast/Reference.ts:26]: this makes sense, will make the change

**Pull request closed without merging**